### PR TITLE
Add `HamiltonianProblem` and its tests

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

File: src/hamiltonian.jl [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+function HamiltonianProblem(H,q0::Real,p0::Real,tspan::Tuple{T,T};kwargs...) where T<:Real
+    @inline function dq(t,x,v,dx)
+        dx[1] = ForwardDiff.derivative(v->H(x[1], v), v[1])
+    end
+    @inline function dp(t,x,v,dv)
+        dv[1] = ForwardDiff.derivative(x->-H(x, v[1]), x[1])
+    end
+
+    return ODEProblem((dq,dp), ([q0],[p0]), tspan; kwargs...)
```
Comment (ChrisRackauckas) [inline src/hamiltonian.jl]: Write this not as vectors. Instead we should just add the out-of-place form for symplectic.

**Comment by Datseris** [comment]
I was going to do it during my holidays, that start tomorrow, but I got destroyed! Good job nontheless, more time to focus on `DynamicalSystems` huhu

**Comment by YingboMa** [comment]
> How can one make it so that the ForwardDiff.gradient! function is called only once?

@Datseris I guess it is not necessary to minimize the number of `ForwardDiff.gradient!` call, since at the end of the day, `ForwardDiff` still needs to take the gradient by chunks anyway.

File: src/hamiltonian.jl [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using OrdinaryDiffEq, ForwardDiff
+
+function HamiltonianProblem(H,q0::Real,p0::Real,tspan::Tuple{T,T};kwargs...) where T<:Real
+    @inline function dq(t,x,v)
+        ForwardDiff.derivative(v->H(x[1], v), v[1])
```
Comment (ChrisRackauckas) [inline src/hamiltonian.jl]: don't need to `first` these

**Pull request merged**